MRG: Add ability to anonymize an entire BIDS dataset

hoechenberger opened the pull request:
Adds `write.anonymize_dataset()`

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [ ] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

hoechenberger commented on mne_bids/write.py:
```diff
@@ -2152,3 +2156,295 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+    datatypes : collection of str | str | None
+        Which data type to anonymize. If can be ``meg``, ``eeg``, ``ieeg``, or
+        ``anat``. Multiple data types may be passed as a collection of strings.
+        If ``None``, try to anonymize the entire input dataset.
```
Resolved in 1451711

agramfort commented on mne_bids/write.py:
```diff
@@ -2158,3 +2162,309 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+        Which data type to anonymize. If can be ``meg``, ``eeg``, ``ieeg``, or
+        ``anat``. Multiple data types may be passed as a collection of strings.
+        If ``None``, try to anonymize the entire input dataset.
+    rng_seed : int | None
```
don't be creative here :) it's called random_state everywhere in mne and use the check_random_state function to get an rng from it.

hoechenberger replied:
Oh. I only found `seed` in the docdict and wasn't happy with it. Will have a look at what you proposed.

agramfort commented on mne_bids/write.py:
```diff
@@ -2152,3 +2156,295 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+        json_out = json.loads(
+            bp_out_json.fpath.read_text(encoding='utf-8')
+        )
+        exclude_keys = (
+            'PatientName',
+            'PatientID',
+            'PatientBirthDate',
+            'PatientSex',
+            'PatientWeight',
+            'AcquisitionTime',
+            'AcquisitionDateTime'
+        )
```
I agree a white list seems more secure...

hoechenberger replied:
We're using a whitelist now: 28248a8

hoechenberger commented on mne_bids/write.py:
```diff
@@ -2158,3 +2163,372 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+    matches = bids_root_in.glob('sub-*/**/sub-*.*')
+    valid_matches = []
+    for f in matches:
+        if (
+            (
+                f.parent.name in requested_datatypes and
+                f.suffix in allowed_extensions and
+                (
+                    (
+                        # Electrophysiological recordings ust have a `task`
+                        # entity
+                        f.stem.endswith(allowed_suffixes_electrophys) and
+                        'task-' in f.name
+                    ) or
+                    (
+                        # stem will contain `.nii` for `.nii.gz` files
+                        any(suffix in f.stem
+                            for suffix in allowed_suffixes_anat)
+                    )
+                )
+            ) or
+            (
+                _check_finecal_path(f) or
+                _check_crosstalk_path(f)
+            )
+        ):
+            valid_matches.append(f)
```
Looks like Lisp! 😅🙈😁 But it does help me.

Can you live with that? 🤔

hoechenberger commented on mne_bids/write.py:
```diff
@@ -2158,3 +2163,340 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+    # Ensure we convert empty-room recordings first, as we'll want to pass
+    # their anonymized path when writing the associated experimental recordings
+    if 'meg' in requested_datatypes:
+        # TODO this heuristic is not great
```
```suggestion
```

hoechenberger commented on mne_bids/write.py:
```diff
@@ -2158,3 +2163,345 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+        f'\n'
+    )
+    if daysback is None:
+        msg += 'Not shifting recording date (found anatomical scans only)'
```
```suggestion
        msg += 'Not shifting recording dates (found anatomical scans only)'
```

hoechenberger commented on mne_bids/config.py:
```diff
@@ -377,3 +378,90 @@ def _map_options(what, key, fro, to):
+    'InstitutionalDepartmentName',
+    'InstitutionAddress',
+    'DeviceSerialNumber',
+    # MRI
```
```suggestion
    # MRI
    # Many of these are not standardized, but produced by dcm2niix.
```

hoechenberger commented:
Rendered example docs:
https://5449-89170358-gh.circle-artifacts.com/0/dev/auto_examples/anonymize_dataset.html

Rendered API docs:
https://5449-89170358-gh.circle-artifacts.com/0/dev/generated/mne_bids.anonymize_dataset.html#mne_bids.anonymize_dataset

agramfort commented on mne_bids/write.py:
```diff
@@ -2158,3 +2163,387 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+        ]
+
+        if subject_mapping == 'auto':
+            participants_out = rng.choice(
+                a=range(100, 1000),
+                replace=False,
+                size=len(participants_in)
+            )
```
I think it's a bit too convoluted. I would generate subject numbers from 1 to n_subjects but use a random permutation. Doing so makes is easy to know how many subjects you have etc. Having a subject number of 332 and you have only one subject suggests it's part of a bigger dataset.

agramfort replied:
to then illustrate this in the example just write first data subjects that initials or subject names and you will see that it will get converted to 01 etc...

hoechenberger replied:
I understand what you're saying but I'm not 100% convinced yet. What I did here is that I basically implemented how we used to produce pseudonymous IDs in my old lab. I always thought 3-digit IDs were a good idea because they allow for lots of "randomness".

I'd be interested to hear what @sappelhoff and @cbrnr think about this?

Also CC @SophieHerbst

cbrnr replied:
I agree with @agramfort that assigning numbers from 1 to N (prepended with as many zeros as necessary) is fine as long as the original subject order is randomly permuted.

hoechenberger replied:
If N < 10, should we prepend a 0 too?

sappelhoff replied:
agreed with Alex and Clemens as well :+1: 

> If N < 10, should we prepend a 0 too?

:+1:

cbrnr replied:
I wouldn't append a 0 if this is not necessary - why would you do that? Data collection is already done, and if there are N < 10 subjects then I'd call them 1...9.

sappelhoff replied:
okay, fair - I misread that "N < 10". Right, if there are no numbers beyond 9, I also see no need for zero-padding ... ordering  will also work without that in that case.

hoechenberger replied:
But we e.g. also always zero-pad runs < 10 … wouldn't that be inconsistent?

sappelhoff replied:
yeah .. why do we do that even? :thinking: 

https://bids-specification.readthedocs.io/en/stable/02-common-principles.html#participant-names-and-other-labels

> In <index>es we RECOMMEND using zero padding (for example, 01 instead of 1 **if you have more than nine subjects**) to make alphabetical sorting more intuitive. 

**emphasis by me**

hoechenberger replied:
ok will not pad then  for N < 10 :)

agramfort reviewed (commented):
just an API suggestion. The rendered doc looks great !

hoechenberger commented on examples/anonymize_dataset.py:
````diff
@@ -0,0 +1,198 @@
+# %%
+# You can see that the subject ID was changed to a number (in this case, the
+# digit ``1```), and the recording dates have been shifted backward in time (as
+# indicated by the ``emptyroom`` session name). Anonymous subject IDs are
````
```suggestion
# indicated by the ``emptyroom`` session name). Anonymized subject IDs are
```

hoechenberger commented:
I've changed the subject ID anonymization in the proposed way.

Rendered API docs:
https://5453-89170358-gh.circle-artifacts.com/0/dev/generated/mne_bids.anonymize_dataset.html

Example docs:
https://5453-89170358-gh.circle-artifacts.com/0/dev/auto_examples/anonymize_dataset.html

hoechenberger commented:
Good to merge from my end.

jasmainak commented on examples/anonymize_dataset.py:
```diff
@@ -0,0 +1,203 @@
+# the timeshift may differ (unless you excplicitly specify them as described
+# above), as they are determined randomly.
+#
+# To ensure results are reproducible across runs, you can pass the
+# ``random_state`` parameter, causing the random number generator to produce
+# the same results every time you execute the function.
```
but then you lose anonymity when the script to anonymize is shared :)

hoechenberger replied:
Only if you share it along with the original dataset … in which case you lose anonymity anyway, no?

jasmainak replied:
fair enough :)

jasmainak commented on mne_bids/write.py:
```diff
@@ -2158,3 +2163,400 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+        they are ``'auto'``.
+    %(verbose)s
+    """
+    from mne_bids import update_sidecar_json  # avoid circular import
```
should go into a utils file?

hoechenberger replied:
What do you mean exactly?

Also I despise `utils` modules, they tend to become a catch-all where one dumps everything

jasmainak replied:
1. ramensoup() and noodles() is in file soup.py and friednoodles() and pepper() is in file maincourse.py
2. ramensoup() and friednoodles() both call pepper() and noodles()

Circular import, right? Solution -> move pepper() and noodles() into pantry.py. Basically circular import indicates pathological organization that needs to be fixed (in most cases ...)

hoechenberger replied:
I'd argue that the entire Python import system is broken, at least when comparing it to modern systems like the one in TypeScript…

I can try to see if a relative import works. Why aren't we using those anyway, @sappelhoff?

hoechenberger replied:
@jasmainak I fixed it by importing from `mne_bids.sidecar_updates`: c8ad05a

Seems one of our issues is that we import ~everything into the `mne_bids` namespace …

sappelhoff replied:
> I can try to see if a relative import works. Why aren't we using those anyway, @sappelhoff?

I find relative imports unreadable and confusing. Also, PEP8 says:

> Absolute imports are recommended, as they are usually more readable and tend to be better behaved (or at least give better error messages) if the import system is incorrectly configured (such as when a directory inside a package ends up on sys.path)

https://www.python.org/dev/peps/pep-0008/

jasmainak replied:
yeah, this looks like the right solution. Don't import everything into `mne_bids` namespace

jasmainak commented on mne_bids/write.py:
```diff
@@ -2158,3 +2163,400 @@ def write_meg_crosstalk(fname, bids_path, verbose=None):
+) -> int:
+    """Try to find a suitable "daysback" for anonymization."""
+    raw = read_raw_bids(bids_path=bids_path, verbose='error')
+    daysback_min, daysback_max = get_anonymization_daysback([raw])
```
using just a single file defeats the purpose of using this function, no? what's min/max ?

hoechenberger replied:
I do this for performance reasons, e.g. for large datasets on an NAS I don't want to have to perform `read_info()` on hundreds or thousands of files just to determine `daysback`. Therefore, I'm picking one recording and chop off 365 days from the reported `min` and `max`, hoping that this will cover all recordings in the dataset.

Not the greatest solution, and ideas for alternative approaches are welcome!

jasmainak replied:
I see, probably won't work for longitudinal datasets but BIDS anonymization is anyway broken for that so I can live with this :)

hoechenberger replied:
cc @agramfort

hoechenberger replied:
After discussion with @agramfort I'll now try to come up with an optimized way to take a larger amount of data into account when estimating a good `daysback`. Will push something shortly.

hoechenberger replied:
My new approach is to pick one run per session for each participant, largely reducing the number of files we need to read in multi-run experiments.

When reading the infos from these files, we display a progress bar if 20 or more files are to be processed. For fewer files, no progress bar is displayed.
(we may wish to adjust this threshold … my idea here was to avoid displaying a progress bar that immediately reaches 100%, cluttering the terminal) Note that this only applies to reading the infos to generate a valid `daysback`. For the actual anonymization, we always display progress bars.

hoechenberger commented:
@agramfort @jasmainak I've addressed all review comments and this is so far good to merge. I will now put this to the test with a real-world dataset and report back asap. But feel free to (re-)review the code, it should be finalized now

agramfort commented on .circleci/config.yml:
```diff
@@ -40,12 +30,6 @@ jobs:
-        - save_cache:
-            key: data-cache-0-{{ checksum "./mne_bids/__init__.py" }}
-            paths:
-              - ~/mne_data
```
I don't think we should merge this right?

hoechenberger replied:
Actually I did remove the caching here on purpose because we used outdated versions of some datasets, causing CI to fail … but I can revert this again and see how it goes

sappelhoff replied:
You should be able to just bump the cache by incrementing the number: data-cache-0... --> data-cache-1...

agramfort commented on examples/anonymize_dataset.py:
```diff
@@ -0,0 +1,203 @@
+# %%
+# Specifying subject IDs
+# ----------------------
+# Anonymized subject IDs are automatically generated as unique 3-digit numbers.
```
this is not true anymore right?

hoechenberger replied:
Good catch, this is fixed now.

jasmainak commented on examples/anonymize_dataset.py:
````diff
@@ -0,0 +1,203 @@
+# You can see that the subject ID was changed to a number (in this case, the
+# digit ``1```), and the recording dates have been shifted backward in time (as
+# indicated by the ``emptyroom`` session name). Anonymized subject IDs are
+# simply a permutation of the set :math:`\{1, 2, \ldots, N\}``, where :math:`N`
+# is the total number of participants (excluding the ``emptyroom``
+# pseudo-subject). Zero-padding will be added to the subject IDs as needed to
+# allow for convenient sorting.
````
```suggestion
# zero padded and numbered from 1 to N, where N
# is the total number of participants
```
drop the math terminology for lay user :)

jasmainak commented on examples/anonymize_dataset.py:
```diff
@@ -0,0 +1,203 @@
+# Limiting to specific data types
+# -------------------------------
+# By default, :func:`mne_bids.anonymize_dataset` will anonymize
+# electrophysiological data and anatomical MR scans (T1-weightes and FLASH).
```
```suggestion
# electrophysiological data and anatomical MR scans (T1-weighted and FLASH).
```

jasmainak commented on examples/anonymize_dataset.py:
```diff
@@ -0,0 +1,203 @@
+
+# %%
+# Reproducibility
+# ---------------
```
while I get the point of reproducibility, are you somehow recommending users to run the function every time the dataset changes? E.g., when a new subject is added. If so, question: what happens to padding in that case? Imagine you had 9 subjects and another subject was added. Now the old subjects will start having padding and it won't be reproducible any more

hoechenberger replied:
No, incremental anonymization is not supported. If the input dataset changes, anonymization should be run on the entire dataset again. We sort-of enforce this by requiring that the output directory do not exist at the time `anonymize_dataset()` is called.

hoechenberger replied:
I have now tried to clarify things in 246907a

jasmainak replied:
looks good!

jasmainak reviewed (approved):
Mostly left some comments on example. I approve if @agramfort approves :)

hoechenberger commented on examples/anonymize_dataset.py:
```diff
@@ -0,0 +1,212 @@
+# .. note::
+#    Passing ``random_state`` merely guarantees that subject IDs and time shift
+#    remain the same across anonymization runs if the original dataset
+#    remained unchanged. It does **not** allow you to incrementally data
```
```suggestion
#    remained unchanged. It does **not** allow you to incrementally add data
```

agramfort reviewed (approved):
+1 for MRG when green !

thx @hoechenberger

agramfort commented:
Thx 🙏 @hoechenberger
